util/collections.py

```diff
@@ -0,0 +1,344 @@
+CONTENT_KEY = 'content'
+MIME_TYPE_KEY = 'mime-type'
+MIME_TYPE_TEXT = 'text/plain'
+CONTENT_TYPE_SPEECH = 'speech'
+CONTENT_TYPE_TRANSCRIPT = 'transcript'
+META = {
+    SCHEMA_KEY: [
+        {CONTENT_KEY: CONTENT_TYPE_SPEECH, MIME_TYPE_KEY: AUDIO_TYPE_OPUS},
+        {CONTENT_KEY: CONTENT_TYPE_TRANSCRIPT, MIME_TYPE_KEY: MIME_TYPE_TEXT}
+    ]
+}
+
+COLUMN_FILENAME = 'wav_filename'
+COLUMN_FILESIZE = 'wav_filesize'
+COLUMN_TRANSCRIPT = 'transcript'
+
+
+class CollectionSample(Sample):
+    def __init__(self, sample_id, audio_type, raw_data, transcript):
+        super().__init__(audio_type, raw_data)
+        self.sample_id = sample_id
+        self.transcript = transcript
+
+
+class DirectSDBWriter:
+    def __init__(self, sdb_filename, buffering=BUFFER_SIZE):
+        self.sdb_filename = sdb_filename
+        self.sdb_file = open(sdb_filename, 'wb', buffering=buffering)
+        self.offsets = []
+        self.num_samples = 0
+
+        self.sdb_file.write(MAGIC)
+
+        meta_data = json.dumps(META).encode()
+        self.write_big_int(len(meta_data))
+        self.sdb_file.write(meta_data)
+
+        self.offset_samples = self.sdb_file.tell()
+        self.sdb_file.seek(2 * BIGINT_SIZE, 1)
+
+    def write_int(self, n):
+        return self.sdb_file.write(n.to_bytes(INT_SIZE, BIG_ENDIAN))
+
+    def write_big_int(self, n):
+        return self.sdb_file.write(n.to_bytes(BIGINT_SIZE, BIG_ENDIAN))
+
+    def __enter__(self):
+        return self
+
+    def add(self, sample):
+        def to_bytes(n):
+            return n.to_bytes(INT_SIZE, BIG_ENDIAN)
+        sample.convert(AUDIO_TYPE_OPUS)
+        opus = sample.audio.getbuffer()
+        opus_len = to_bytes(len(opus))
+        transcript = sample.transcript.encode()
+        transcript_len = to_bytes(len(transcript))
+        entry_len = to_bytes(len(opus_len) + len(opus) + len(transcript_len) + len(transcript))
+        buffer = b''.join([entry_len, opus_len, opus, transcript_len, transcript])
+        self.offsets.append(self.sdb_file.tell())
+        self.sdb_file.write(buffer)
+        self.num_samples += 1
+
+    def close(self):
+        if self.sdb_file is None:
+            return
+        offset_index = self.sdb_file.tell()
+        self.sdb_file.seek(self.offset_samples)
+        self.write_big_int(offset_index - self.offset_samples - BIGINT_SIZE)
+        self.write_big_int(self.num_samples)
+
+        self.sdb_file.seek(offset_index + BIGINT_SIZE)
+        self.write_big_int(self.num_samples)
+        for offset in self.offsets:
+            self.write_big_int(offset)
+        offset_end = self.sdb_file.tell()
+        self.sdb_file.seek(offset_index)
+        self.write_big_int(offset_end - offset_index - BIGINT_SIZE)
+        self.sdb_file.close()
+
+    def __len__(self):
+        return len(self.offsets)
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.close()
+
+
+class SortingSDBWriter:  # pylint: disable=too-many-instance-attributes
```

Comment:
It does external sorting - which is required as soon as files get huge (which is the primary reason for using SDBs).

Reply:
I don't fully understand why external sorting is required. You're only ever writing a new sorted file from scratch, correct? There's no utility to take an unsorted SDB file and sort it. And the SDB file has an index with all the files and their sizes. So to generated a sorted SDB out of several sources, can't you just merge all of the individual indices, sort the index in memory, then write the output SDB directly in order without any sorting? I feel like I'm missing some key aspect here.

Reply:
Until now the utility is `build_sdb.py`.

The only reason for external sorting is performance:
Imagine an index merging approach. It would:
- seek within the file for every single sample it reads (no big issue on SSDs, but:)
- this renders any read-ahead buffering useless (major performance advantage gone)
- which is particularly worse in case of a remote FS where all OPs have to travel forth and back through network

At least DSAlign will have to have this code (as it will have to produce samples in a source-file dependent order without knowing the full index up front - "open mp3, extract samples and distribute them across sub-sets/partitions"). I thought it is a good feature to have. It's also not that much more "expensive" than an index based approach, which would have to be implemented as another solution for something that has already been solved.

Reply:
I understand the need to have the data in the order that it will be read to minimize seeks. What I'm talking about is sorting the list of files by the file size (an index), and then writing the SDB in the correct seek minimizing order by just following the order of the sorted index and writing the files. They will still be in order. The list of files can be sorted in memory, so you don't have to use external sorting.

Reply:
This is only true if all source files are CSVs.

Reply:
Why? Doesn't the SDB have an index in it with the list of files and their sizes? You can take that list, merge with CSVs or other SDBs, and then sort the entire list.

Reply:
Yeah - but at some point (knowing your indices) you have to read the contents of a potentially unsorted source SDB and seek around.
This code covers an edge case, yes. But it is already there (for other reasons). So why implementing something that already works a second time?

Reply:
Because this is a bunch of complicated code that we have to understand, maintain, and will get in the way of contributors. I've already read through this file two times and I'm still feeling a bit lost. I'm trying to understand if the benefits are worth the costs.

So the edge case you're optimizing for is: you have a bunch of sources, and some of them are unsorted SDBs. When building the new sorted SDB with the index approach, you'd have to seek around the unsorted SDB. My proposal: enforce all SDBs are sorted, since that's what we're actually using for DeepSpeech. Do we have any compelling use case for unsorted SDBs?

Reply:
OK - I'll remove all sorting now and let `build_sdb.py` state that - for getting a sorted result SDB - one has to ensure that all sources have to be already sorted (which is automatically given for CSV files, as they are sorted upon reading). I'd rather not implement sorting on base of index differences in SDBs, as this not guaranteeing correctness (compression).